### PR TITLE
Phase 6: Iterative Weight-Tied Transolver — Deeper via Block Reuse

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current Transolver uses 3 distinct blocks (`n_layers=3`): 2 intermediate blocks (different weights) + 1 final block. A key question is whether more computational depth would help — but simply adding more distinct blocks increases parameter count and risks overfitting.

**Weight-tied iteration** gives depth without more parameters: replace the 2 intermediate blocks with a SINGLE shared block run K times. This is inspired by Deep Equilibrium Networks (Bai et al. NeurIPS 2019) and Unrolled Optimization, applied here as a lightweight architectural change.

**Why it might help:** p_tan = 29.1 is 2.5x worse than p_in = 12.2. The tandem aft foil operates in the fore foil's wake — a complex long-range interaction. The hypothesis is this is a **receptive-field problem**: the 3-block architecture has limited range for fore→aft signal propagation. Running the same block K=3 or K=4 times effectively gives more computational steps for these long-range signals to propagate, without the parameter bloat of adding distinct blocks.

**Physical analogy:** RANS pressure-velocity coupling is itself an iterative solver (SIMPLE/PISO). Giving the model explicit iterative structure aligns its computational graph with the iterative nature of the underlying physics.

**Note on model scale-up (frieren #2100):** That PR tests deeper/wider with DISTINCT blocks (more unique parameters). This PR tests deeper with SHARED blocks (same or fewer parameters). The mechanisms are different — if scale-up fails, weight tying should still be tried; if scale-up succeeds, weight tying may further amplify it.

**Key paper:** Bai et al., "Deep Equilibrium Models", NeurIPS 2019. The unrolled K-iteration version (no Jacobian solver): Fung et al., "Fixed Point Networks", 2021.

### Code architecture (read this carefully before coding)

Looking at `train.py`, the Transolver has:
- `self.blocks`: a `ModuleList` of `n_layers` blocks
- In `forward`, lines ~790-808: loops over `self.blocks[:-1]` then calls `self.blocks[-1]` separately (the last block has `last_layer=True` and unique features: `field_decoder`, `adaln_output`, `domain_velhead`, `pressure_first`, `pressure_deep`)
- The intermediate blocks (`last_layer=False`) are identical in architecture — only their weights differ

### 1. Add new config flag (in Config dataclass, ~line 940)

```python
n_iter: int = 0  # if > 0, use weight-tied iterative intermediate blocks (0 = standard n_layers)
```

### 2. Modify Transolver.__init__ block creation (~lines 645-676)

Add a conditional after the existing block creation:

```python
# AFTER self.blocks = nn.ModuleList([...]) at line 645:
if cfg.n_iter > 0:
    # Weight-tied mode: keep only 2 blocks (shared intermediate + last)
    # blocks[0] = shared intermediate block (run n_iter times)
    # blocks[1] = last block (run once, retains all last_layer features)
    assert len(self.blocks) >= 2, "n_iter requires at least 2 blocks"
    shared_block = self.blocks[0]   # keep first block as the shared intermediate
    last_block = self.blocks[-1]    # keep last block as-is
    self.blocks = nn.ModuleList([shared_block, last_block])  # discard unused blocks
self.n_iter = cfg.n_iter if cfg.n_iter > 0 else len(self.blocks) - 1
```

### 3. Modify Transolver.forward block loop (~lines 790-812)

Replace the existing block loop with:

```python
# Instead of:
#   for block in self.blocks[:-1]:
#       fx = block(fx, ...)
# Use:
for _ in range(self.n_iter):
    fx = self.blocks[0](fx, raw_xy=raw_xy, tandem_mask=is_tandem,
                         condition=block_condition, zone_features=zone_features)
# Last block (unchanged):
fx = self.blocks[-1](fx, raw_xy=raw_xy, tandem_mask=is_tandem,
                      condition=last_condition, zone_features=zone_features)
```

**Note:** Make sure both the main path AND the `pressure_deep` path (if it has a separate block loop) are updated. Search for `self.blocks[:-1]` and update all occurrences.

### 4. Experiment matrix (8 GPUs)

| GPU | Config | Seed | Notes |
|-----|--------|------|-------|
| 0 | Baseline (n_iter=0, 3 distinct blocks) | 42 | Control |
| 1 | Baseline (n_iter=0, 3 distinct blocks) | 43 | Control |
| 2 | `--n_iter 2` (2 shared iters + 1 last) | 42 | Same depth, weight-tied |
| 3 | `--n_iter 2` (2 shared iters + 1 last) | 43 | Same depth, weight-tied |
| 4 | `--n_iter 3` (3 shared iters + 1 last) | 42 | +1 depth, weight-tied |
| 5 | `--n_iter 3` (3 shared iters + 1 last) | 43 | +1 depth, weight-tied |
| 6 | `--n_iter 4` (4 shared iters + 1 last) | 42 | +2 depth, weight-tied |
| 7 | `--n_iter 4` (4 shared iters + 1 last) | 43 | +2 depth, weight-tied |

Use `--wandb_group phase6/iterative-transolver` for all runs. Pass `--n_layers 3` for all runs to ensure consistent block construction.

**Base command:**
```bash
python train.py --agent thorfinn --wandb_group phase6/iterative-transolver \
  --wandb_name "thorfinn/iter-k${k}-s${seed}" \
  --n_iter ${k} \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed ${seed}
```

Baseline (n_iter=0): same command without `--n_iter`.

### Important notes
- DO NOT modify MAX_EPOCHS (500) or MAX_TIMEOUT (180.0) in train.py
- `--n_layers 3` is required to ensure 3 blocks are created before the weight-tying collapses them to 2
- Search for ALL occurrences of `self.blocks[:-1]` in forward() — there may be multiple (main path + optional pressure_deep path). Update all of them to use the `n_iter` loop.
- With `n_iter=2`, the model has slightly fewer parameters than baseline (blocks[1] is discarded). This is expected.
- Watch for gradient explosion with n_iter=4 — more unrolled steps = longer gradient path. The existing gradient clipping in train.py should handle this, but watch the grad_norm metric in W&B.
- Log `n_iter` and `n_params` in wandb at the start of training so we can compare

## Baseline

Current best (16-seed ensemble, PR #2093):
- p_in: **12.1** | p_oodc: **6.6** | p_tan: **29.1** | p_re: **5.8**

Single-model baseline (8-seed mean):
- p_in: **13.03** | p_oodc: **7.83** | p_tan: **30.29** | p_re: **6.45**

## Reporting

Post results with:
- Individual metrics for all 8 runs
- Mean ± std per n_iter vs baseline
- Parameter counts for each config (to confirm weight sharing works)
- Gradient norms (did any runs show instability?)
- W&B run IDs

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Summary: Weight-tied iterative blocks **hurt performance** — n_iter=2 mildly worse, n_iter=3/4 significantly worse.

### Individual Run Metrics

| GPU | Config | Seed | p_in | p_oodc | p_tan | p_re | val_loss | Best Epoch | Params |
|-----|--------|------|------|--------|-------|------|----------|------------|--------|
| 0 | **Baseline** (3 distinct blocks) | 42 | **12.8** | **7.7** | **30.1** | **6.4** | 0.386 | 156 | 1,810,787 |
| 1 | **Baseline** (3 distinct blocks) | 43 | **13.0** | **8.0** | **30.0** | **6.6** | 0.389 | 157 | 1,810,787 |
| 2 | n_iter=2 (same depth, weight-tied) | 42 | 13.1 | 7.8 | 31.4 | 6.5 | 0.397 | 156 | 1,500,585 |
| 3 | n_iter=2 (same depth, weight-tied) | 43 | 14.3 | 8.2 | 30.8 | 6.7 | 0.398 | 156 | 1,500,585 |
| 4 | n_iter=3 (+1 depth, weight-tied) | 42 | 14.2 | 8.8 | 30.6 | 7.1 | 0.411 | 128 | 1,500,585 |
| 5 | n_iter=3 (+1 depth, weight-tied) | 43 | 16.6 | 9.1 | 32.1 | 7.1 | 0.423 | 127 | 1,500,585 |
| 6 | n_iter=4 (+2 depth, weight-tied) | 42 | 18.6 | 10.0 | 32.9 | 8.1 | 0.452 | 107 | 1,500,585 |
| 7 | n_iter=4 (+2 depth, weight-tied) | 43 | 18.4 | 11.0 | 30.9 | 7.9 | 0.451 | 107 | 1,500,585 |

### Group Comparison

| Config | p_in | p_oodc | p_tan | p_re | Epochs | Params |
|--------|------|--------|-------|------|--------|--------|
| **Baseline** | **12.9** | **7.8** | **30.1** | **6.5** | 156-157 | 1.81M |
| n_iter=2 | 13.7 (+6%) | 8.0 (+2%) | 31.1 (+3%) | 6.6 (+2%) | 156 | 1.50M |
| n_iter=3 | 15.4 (+19%) | 8.9 (+14%) | 31.4 (+4%) | 7.1 (+9%) | 127-128 | 1.50M |
| n_iter=4 | 18.5 (+43%) | 10.5 (+34%) | 31.9 (+6%) | 8.0 (+23%) | 107 | 1.50M |

### Analysis

**Two confounding factors drive the degradation:**

1. **Fewer parameters (~17% fewer):** Weight-tying discards the second intermediate block, reducing from 1.81M to 1.50M params. The n_iter=2 case is the cleanest comparison — same computational depth (2 intermediate passes + 1 final), but with shared weights. It's already 2-6% worse, indicating that the distinct intermediate blocks learn complementary representations.

2. **Fewer training epochs (for n_iter≥3):** More iterations per forward pass = slower per-epoch = fewer epochs in the 180-min budget. n_iter=3 only reaches epoch 127-128 vs 156 for baseline. n_iter=4 only reaches epoch 107. This means the cosine LR schedule barely completes (T_max=160), and the model doesn't converge as well.

**Key insight from n_iter=2:** Even with identical computational depth and training epochs, weight sharing hurts by ~2-6%. This tells us the two intermediate blocks learn meaningfully different representations — they're not just iterating the same computation. The "iterative solver" analogy doesn't hold for this architecture.

**n_iter=4 worst metric: p_in=18.5 (+43%).** The degradation on in-distribution samples is most severe, suggesting the shared block cannot specialize its representation for the different depths.

### Training Details

- W&B group: `phase6/iterative-transolver`
- All runs hit the 180-min wall-clock limit
- No gradient instability observed (gradient clipping was sufficient)

### W&B Run IDs

| GPU | Config | Run ID |
|-----|--------|--------|
| 0 | Baseline s42 | onmaryjt |
| 1 | Baseline s43 | xhpmloz0 |
| 2 | n_iter=2 s42 | i8xsa2g0 |
| 3 | n_iter=2 s43 | 9ny3ctv2 |
| 4 | n_iter=3 s42 | 55m69afg |
| 5 | n_iter=3 s43 | i6ucjvki |
| 6 | n_iter=4 s42 | i048aror |
| 7 | n_iter=4 s43 | 1dxezufa |

### Command Used

```bash
# Baseline (GPUs 0-1)
CUDA_VISIBLE_DEVICES=$gpu python train.py --agent thorfinn \
  --wandb_group phase6/iterative-transolver --wandb_name "thorfinn/baseline-s${seed}" \
  --asinh_pressure --asinh_scale 0.75 [all baseline flags] --seed ${seed}

# Weight-tied (GPUs 2-7, varying n_iter)
CUDA_VISIBLE_DEVICES=$gpu python train.py --agent thorfinn \
  --wandb_group phase6/iterative-transolver --wandb_name "thorfinn/iter-k${k}-s${seed}" \
  --n_iter ${k} --asinh_pressure --asinh_scale 0.75 [all baseline flags] --seed ${seed}
```

Peak memory: ~38 GB (baseline), slightly less for weight-tied (fewer params).

### Suggested Follow-ups

- **This hypothesis is negative.** Weight-tied blocks consistently degrade performance. The two intermediate blocks in the baseline learn distinct representations that cannot be replicated by running one block multiple times.
- If depth is the goal, **adding more distinct blocks** (as explored in frieren's #2100 model scale-up) is the right approach, not weight sharing.
- The only salvageable variant might be n_iter=2 with a warm-start: initialize the shared block from the average of the two baseline intermediate blocks. But the improvement would be marginal at best.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2103 because Closing — weight-tied iterations consistently degrade performance. The degradation is monotonic with iteration count:

| Config | p_in (Δ) | p_oodc (Δ) | p_tan (Δ) | p_re (Δ) |
|--------|---------|-----------|---------|---------|
| Baseline (3 distinct blocks) | 12.9 | 7.8 | 30.1 | 6.5 |
| n_iter=2 (same depth, shared) | 13.7 (+6%) | 8.0 (+2%) | 31.1 (+3%) | 6.6 (+2%) |
| n_iter=3 | 15.4 (+19%) | 8.9 (+14%) | 31.4 (+4%) | 7.1 (+9%) |
| n_iter=4 | 18.5 (+43%) | 10.5 (+34%) | 31.9 (+6%) | 8.0 (+23%) |

**Key insight (from n_iter=2):** Even at identical computational depth and training epochs, weight sharing hurts 2-6%. This proves the two intermediate blocks learn *complementary* representations that cannot be replicated by running one block multiple times. The 'iterative solver' analogy from DEQ doesn't hold here — the Transolver blocks are not approximating fixed-point iteration.

Combined with #2100 (scale-up): distinct blocks are important, but adding MORE distinct blocks doesn't help either. The 3-block architecture is the right structural choice. Well-executed experiment with a clean experimental design (same-epoch n_iter=2 as a controlled comparison)..
